File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

unordered_map<string, vector<string>> graph;
unordered_map<string, int> visited;
int total;
bool finish = false;
vector<string> answer;

void bt(string curr){
    
    answer.push_back(curr);
    
    if (answer.size() == total + 1){
        finish = true;
        return;
    }
    
    for(string& next : graph[curr]){
        string path = curr + next;
        if (visited[path] <= 0) continue;
        visited[path] -= 1;
        bt(next);
        if (finish) return;
        visited[path] += 1;
    }
    answer.pop_back();
    
}


vector<string> solution(vector<vector<string>> tickets) {
    total = tickets.size();
    
    for(auto& ticket : tickets){
        graph[ticket[0]].push_back(ticket[1]);
        string path = ticket[0] + ticket[1];
        if (!visited.count(path)) visited[path] = 0;
        visited[path] += 1;
    }
    for (auto& nodes : graph){
        sort(nodes.second.begin(), nodes.second.end());
    }
    
    // // 0. 출발점 정하기 -> 사전순으로 빠른 곳부터 출발
    // vector<string> order;
    // for (auto& p : graph){
    //     order.push_back(p.first);
    // }
    // sort(order.begin(), order.end());

    bt("ICN");
    return answer;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    
    static Map<String, Integer> cnt;
    static Map<String, List<String>> graph;
    
    static List<String> answer;
    static List<String> curr;
    static int total;
    
    boolean bt(String node, int used){
        boolean rnt = false;
        curr.add(node);
        if (used == total){
            answer = new ArrayList<>(curr);
            return true;
        }
        else{
            for (String next : graph.getOrDefault(node, new ArrayList<String>())){
                
                if (rnt == true) break;
                
                String move = node + " " + next;
                if (cnt.getOrDefault(move, 0) <= 0) continue;
                
                cnt.put(move, cnt.get(move) - 1);
                rnt = bt(next, used+1);
                cnt.put(move, cnt.get(move) + 1);
            }
        }        
        curr.remove(curr.size() - 1);
        return rnt;
    }
    
    
    public String[] solution(String[][] tickets) {
        
        cnt = new HashMap<>();
        graph = new HashMap<>();
        answer = new ArrayList<>();
        curr = new ArrayList<>();
        total = 0;
        
        for (var t : tickets){
            String s = t[0];
            String e = t[1];
            
            // 티켓 개수 설정
            String name = s + " " + e;
            cnt.put(name, cnt.getOrDefault(name, 0) + 1);
            total++;
            
            // graph 설정
            List<String> v = graph.getOrDefault(s, new ArrayList<String>());
            v.add(e);
            graph.put(s, v);
        }
        
        // 내부 노드 정렬
        List<String> startNode = new ArrayList<>();
        
        for(var p : graph.entrySet()){
            var v = p.getValue();
            Collections.sort(v);
        }
        bt("ICN",0);
        
        return answer.toArray(new String[0]);
    }
}